### PR TITLE
Implementing "<test_binary> --list --format json" for use by IDE test explorers / runners

### LINKED ISSUE #107307: Implementing "<test_binary> --list --format json" for use by IDE test explorers / runners
[consider the default output of ```cargo new a --lib```]

## Issue
Test explorers in editors (e.g. VSCode, see pic below) are unable to navigate to test location in the editor from the test explorer.

## Cause
```cargo test --list -q``` used by the editors to get list of tests do not have the source and line / column info.

## Description
Currently ```cargo test --list -q``` shows the following output
```
tests::it_works: test
```

This is used by editor extensions e.g. [vscode rust test adapter](https://marketplace.visualstudio.com/items?itemName=swellaby.vscode-rust-test-adapter) to show & run tests from the tests pane in VS Code. e.g.
image

While it is possible to run and debug the tests from the test explorer, it is not possible to navigate back to the test location as output above is missing that information. E.g. you cannot double click on a test on the test explorer to navigate to its definition in the editor.

Suggestion is to add the source info to the output of --list as follows
```
tests::it_works: test,<path>|<start_line>|<start_col>|<end_line>|<end_col>
```
## Fix proposal
Proposal for fix (& I only have a plan A, happy to receive further guidance) is to update ```test::TestDescAndFn``` as follows

```rs
// https://github.com/rust-lang/rust/blob/master/library/test/src/types.rs#L118
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct TestDesc {
    pub name: TestName,
    pub ignore: bool,
    pub ignore_message: Option<&'static str>,
    pub should_panic: options::ShouldPanic,
    pub compile_fail: bool,
    pub no_run: bool,
    pub test_type: TestType,
    pub source_path: &'static str,     // <- Added
    pub start_line: i32,               // <- Added
    pub start_column: i32,             // <- Added
    pub end_line: i32,                 // <- Added
    pub end_column: i32,               // <- Added
}
```
Once this is available, the corresponding changes can be made to [compiler/rustc_builtin_macros/src/test.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/test.rs#L251) to emit the above info and then in [library/test/src/console.rs](https://github.com/rust-lang/rust/blob/master/library/test/src/console.rs#L166) to write it out.

### PR DESCRIPTION
**Pull request by parthopdas**
Fixes #107307 

PR 1 of 2 - wiring up just the new information + implement the command line changes i.e. --format json + tests

upcoming:
PR 2 of 2 - clean up "#[cfg(not(bootstrap))]" from PR 1

As per the discussions on
- MCP: https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Implementing.20.22.3Ctest_binary.3E.20--list.20--form.E2.80.A6.20compiler-team.23592/near/328747548
- preRFC: https://internals.rust-lang.org/t/pre-rfc-implementing-test-binary-list-format-json-for-use-by-ide-test-explorers-runners/18308
- FYI on Discord: https://discord.com/channels/442252698964721669/459149169546887178/1075581549409484820

### CODE REVIEW

**Comment by parthopdas** [comment]
@Manishearth fyi

**Comment by Manishearth** [comment]
@cuviper For background: The plan here only affects a preexisting unstable CLI option (`--format`), so it should be possible to make these changes without worrying about backcompat. The author is planning on also making a proper RFC for the feature after this lands based on the existing implementation.

(though the TestDescAndFn changes are going to be a pain for nightly compiletest, oh well)

**Comment by Manishearth** [comment]
Also, I do not know enough about TestDescAndFn to have a good opinion on whether this is the right way to make this change, @oli-obk or @cuviper might.

**Comment by parthopdas** [comment]
> (though the TestDescAndFn changes are going to be a pain for nightly compiletest, oh well)
@Manishearth this PR already contains changes to a bunch of tests. but if your comment refers to something else i am happy to fix them if you let me know how to find them.

File: compiler/rustc_span/src/source_map.rs
Comment (bjorn3) [inline compiler/rustc_span/src/source_map.rs:465]: *location

File: compiler/rustc_builtin_macros/src/test.rs
Comment (bjorn3) [inline compiler/rustc_builtin_macros/src/test.rs:294]: Is the end location really useful?

File: compiler/rustc_span/src/source_map.rs
Comment (bjorn3) [inline compiler/rustc_span/src/source_map.rs:466]: What is the `self.files.borrow().source_files.is_empty()` condition necessary for?

File: compiler/rustc_builtin_macros/src/test.rs
Comment (parthopdas) [reply compiler/rustc_builtin_macros/src/test.rs:234]: this is a good suggestion, i think.

sess is not available in source_span, so i added it on test.rs line 234 (the one you highlighted above). however there "self.mir" is not available. any ideas? sess here is ExtCtxt while in debuginfo.rs self is FunctionCtx.

Comment (bjorn3) [reply compiler/rustc_builtin_macros/src/test.rs]: Good question. Didn't notice the `self.mir`. I'm not sure how to replace it.

**Comment by Manishearth** [comment]
note on commit messages: there's no need to put the issue number at the beginning of each message, and you probably don't want commits to have the same message: have each one describe what it does

(right now when looking at a list of commits you don't really see anything useful because they're too long and overlapping)

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [reply compiler/rustc_span/src/source_map.rs:466]: this has not been answered yet.

**Comment by oli-obk** [comment]
Just for visibility, this should get an MCP, I don't see it being controversial, but compiler folks should know about it via that mechanism.

**Comment by parthopdas** [comment]
> (right now when looking at a list of commits you don't really see anything useful because they're too long and overlapping)

oh but they do - just need to expand the description

image

**Comment by parthopdas** [comment]
drats! closed the PR by mistake. @#!$@$!!!

**Comment by Manishearth** [comment]
> oh but they do - just need to expand the description

Yeah, I mean there's no need for the commit header to be the PR title for each one; just have your short message be the main commit message

**Comment by parthopdas** [comment]
> Just for visibility, this should get an MCP, I don't see it being controversial, but compiler folks should know about it via that mechanism.

@oli-obk please point me to the process - happy to do that.

File: compiler/rustc_builtin_macros/src/test.rs
Comment (parthopdas) [reply compiler/rustc_builtin_macros/src/test.rs]: i based my changes on ```span_to_string``` in source_map.rs. also searched unsuccessfully for a way to do what you recommended in there and in tests.rs.

for now i'll close it unless someone can help out.

File: compiler/rustc_span/src/source_map.rs
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs]: fixed.

ps: pretty sure responded to this. not sure why my comments got reset.

File: compiler/rustc_builtin_macros/src/test.rs
Comment (parthopdas) [reply compiler/rustc_builtin_macros/src/test.rs:294]: it is my guess right now, i think i will be needing this for the UI adornments e.g. highlighting the symbol when clicked on the test explorer and adding code lens related glyphs.

ps: pretty sure responded to this. not sure why my comments got reset.

File: compiler/rustc_span/src/source_map.rs
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs:466]: i based it on the following. my guess is it handles the temporary files generated by the compiler.

https://github.com/rust-lang/rust/blob/9aa5c24b7d763fb98d998819571128ff2eb8a3ca/compiler/rustc_span/src/source_map.rs#L441-L448

ps: pretty sure responded to this. not sure why my comments got reset.

**Comment by parthopdas** [comment]
@oli-obk @bjorn3 i am pretty sure i responded to all the reviews in the morning but for some reason they are missing. responded to them again just now. let me know what you think.

File: compiler/rustc_builtin_macros/src/test.rs
Comment (bjorn3) [reply compiler/rustc_builtin_macros/src/test.rs:294]: I see

**Comment by parthopdas** [comment]
@oli-obk raised the MCP https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Implementing.20.22.3Ctest_binary.3E.20--list.20--form.E2.80.A6.20compiler-team.23592/near/328747548

**Comment by parthopdas** [comment]
@oli-obk please can you approve this PR?

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [reply compiler/rustc_span/src/source_map.rs:466]: I think `span_to_string` should call `span_to_location_info` to get the information it needs. This avoids the duplication of all the logic.

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:473]: Maybe return `Lrc<SourceFile>` instead of a symbol, so the caller can decide what to do with it

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: tests/ui/test-attrs/tests-listing-format-json.rs
Comment (oli-obk) [inline tests/ui/test-attrs/tests-listing-format-json.rs:3]: Please add a test showing that without -Zunstablre-opts we don't allow this.

File: compiler/rustc_span/src/source_map.rs
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs]: done

File: tests/ui/test-attrs/tests-listing-format-json.rs
Comment (parthopdas) [reply tests/ui/test-attrs/tests-listing-format-json.rs:3]: done

File: compiler/rustc_span/src/source_map.rs
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs]: > Please rebase the merge commits away and squash your fixup commits (or just make one big commit, that's also ok).

done

> Do we have documentation somewhere for the existing json output of the test suite?

none, the json was implemented without an RFC and through a tracking issue #49359. i will be creating the combined RFC for all of this and we'll have the documentation there. till then the agreement is to leave it under the unstable_opts flag.

**Comment by cuviper** [comment]
I don't see any problems on the library side - deferring to Oli for the whole review.

r? @oli-obk 

>> Please rebase the merge commits away and squash your fixup commits (or just make one big commit, that's also ok).
>
> done

It does not appear so -- in fact you pushed a new `Merge branch 'rust-lang:master' into master`.

**Comment by parthopdas** [comment]
> It does not appear so -- in fact you pushed a new `Merge branch 'rust-lang:master' into master`.

@cuviper i did a pull rebase. i assume we dont mean recreating a new fork and then re-applying all the commits one by one?

**Comment by cuviper** [comment]
I would use `git rebase -i`, then choose which commits to `pick` as-is or `fixup` the prior commit.

**Comment by parthopdas** [comment]
> I would use `git rebase -i`, then choose which commits to `pick` as-is or `fixup` the prior commit.

@cuviper not sure how to do that. thing is this is a 15ish day old PR on my fork and there have been multiple merges syncs etc.
can you point me to some docs / steps?

i started with creating a completely new fork and copy paste but that will lose the conversation in this PR.

**Comment by cuviper** [comment]
You don't need a new fork -- this is something you do on your existing branch. Maybe one of these will help:
https://docs.github.com/en/get-started/using-git/about-git-rebase
https://docs.gitlab.com/ee/topics/git/git_rebase.html

**Comment by parthopdas** [comment]
ok attempted twice: with ```git rebase -i HEAD~200``` and then with ```git rebase -i origin/master~200 master``` - both times large number of conflicts that are unrelated. perhaps there is a way to do it successfully and i do not know about it.

i give up, creating new fork and new PR. i will close this one when the other is ready to go checks passed and stuff.

till then please add your comments to this pr.

**Comment by cuviper** [comment]
From your merge commit, it looks like you have this repo under the name "rust-lang", but you can confirm that with `git remote -v` to see what you have.

While you're on the branch you want to rebase (looks like your local `master`), run `git rebase -i rust-lang/master`.

**Comment by parthopdas** [comment]
it my personal fork for rust-lang/rust - this is what i ses
image

**Comment by parthopdas** [comment]
ok started - https://github.com/rust-lang/rust/pull/108719

**Comment by parthopdas** [comment]
ok https://github.com/rust-lang/rust/pull/108719 is ready for us to move over. all comments from this PR are included in https://github.com/rust-lang/rust/pull/108719.

**Comment by oli-obk** [comment]
You can add the rust-lang repo as a remote with

```
git remote add upstream https://github.com/rust-lang/rust.git
```

Then you can rebase over the latest master branch with

```
git fetch upstream
git rebase -i upstream/master
```

**Comment by parthopdas** [comment]
that was easy enough. thanks @oli-obk. squashed all into 1 commit.

File: compiler/rustc_span/src/source_map.rs
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs:470]: this is also done.

**Comment by parthopdas** [comment]
@oli-obk i have taken in all the comments. please can you approve it or let me know if there are any other comments?

**Comment by oli-obk** [comment]
I will review some time next week

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:450]:
```suggestion
            None => return "no-location".to_string(),
```

otherwise we get `no-location:0:0: 0:0`

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:454]: now that everything is stored in variables, you can inline it

```suggestion
            "{file_name}:{lo_line}:{lo_col}{}",
```

File: compiler/rustc_span/src/source_map.rs
Comment (oli-obk) [inline compiler/rustc_span/src/source_map.rs:461]:
```suggestion
                format!(": {hi_line}:{hi_col}")
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
just a style nit and a missing `return`, then this lgtm

**Comment by parthopdas** [comment]
i'll do the rebase thing as soon as the validation passes.

**Comment by parthopdas** [comment]
@oli-obk took in your suggestions and rebased the PR.

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
Thanks!

@bors r+

**Comment by matthiaskrgr** [comment]
@bors r- failed in a rollup
https://github.com/rust-lang/rust/pull/108982
https://github.com/rust-lang-ci/rust/actions/runs/4385555922/jobs/7678439565#step:26:3680

**Comment by parthopdas** [comment]
@matthiaskrgr @oli-obk @cuviper i get the following failure in the pre-merge CI but not in the validation CI.

please can you provide a hint on how to fix it?
i searched for other pp files with paths but didn't find any + the other stdout file has paths with the CI auto replaced with $DIR.

any hint will be much appreciated, while i search around the code base.

image

**Comment by parthopdas** [comment]
@oli-obk need a mini-review. once you are ok with this. 

i'll rebase as soon as validations are complete

File: compiler/rustc_builtin_macros/src/test.rs
Comment (parthopdas) [inline compiler/rustc_builtin_macros/src/test.rs:385]: @oli-obk the fix was to change ::Local to ::Remapped and then add ```--remap-path-prefix={{src-base}}=/the/src``` in tests-are-sorted.rs

**Comment by oli-obk** [comment]
Please refrain from pinging me repeatedly. I will get to this PR when I have review time.

**Comment by parthopdas** [comment]
sure and my apologies - still learning how things work around here. the logic i applied was lack of ack likely implies request went to bottom of notification stack and was probably missed.

**Comment by oli-obk** [comment]
you can use

@rustbot review

to apply the `waiting-for-review` label, that is independent of notifications and everyone can look at the PRs assigned to themselves that have this label.

**Comment by parthopdas** [comment]
the CI is really good. i love it!

**Comment by oli-obk** [comment]
Curious. UI tests automatically handle slashes, but pretty tests don't. Oh well, seems fine for now

@bors r+

**Comment by parthopdas** [comment]
happy to fix the root cause i.e. transfer the slash handling thingy. just need a hint where to look for. from a cursory search across the code base - could not locate it.

**Comment by oli-obk** [comment]
looks like you also need `// normalize-stderr-test` comment that normalizes away the `fake-test-src-base/test-attrs/` to `$DIR`. Look at other uses of `normalize-stderr-test` to see how that is done.

**Comment by parthopdas** [comment]
so it appears fake-test-src-base is from the test runner. also i am on windows and it appears to handle the path sep char nicely.

hence not messing with it.

https://github.com/rust-lang/rust/blob/e4b9f86054c581d931f8bad0c87523c53e1e8e3f/src/ci/docker/scripts/fuchsia-test-runner.py#L906-L911

**Comment by parthopdas** [comment]
last suite you'll ever wear... again! lol

**Comment by ehuss** [comment]
@bors retry
wasm SIGSEGV

**Comment by parthopdas** [comment]
bors CI failure, tree closed now. updates here https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/segfaults/near/342843524

**Pull request merged**
